File: cracking-the-coding-interview/1-arrays-and-strings.js/1.1-is-unique.js
'use strict';

const expect = require('chai').expect;

/*
Is Unique:

Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?
*/

function isUnique(str) {
  var uniqueMap = new Map();

  for (let i = 0; i < str.length; i++) {
    if (uniqueMap.has(str.charAt(i))) {
      return false;
    }
    else {
      uniqueMap.set(str.charAt(i), 1);
    }
  }
  return true;
}

describe('1.1 Is Unique | Cracking the Coding Interview | Chapter 1 - Arrays and Strings', function() {
  describe('input is empty string', function() {
    it('should return true', done => {
      expect(isUnique('')).to.equal(true);
      done();
    })
  })

  describe('Input: "abcdefg"', function() {
    it('should return true', done => {
      expect(isUnique('abcdefg')).to.equal(true);
      done();
    })
  })

  describe('Input: "abcdegg"', function() {
    it('should return false', done => {
      expect(isUnique('abcdegg')).to.equal(false);
      done();
    })
  })
})

File: cracking-the-coding-interview/01-arrays-and-strings.js/1.1/1.1-is-unique.java
/*
Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

You should first ask your interviewer if the string is an ASCII string or a Unicode String. Asking this question will show an eye for detail and a solid foundation in computer science. We'll assume for simplicity the character set is ASCII. If this assumption is not valid, we would need to increase the storage size.

One solution is to create an array of boolean values, where the flag at index i indicates whether character i is in the alphabet contained in the string. The second time you see this character you can immediately return false.

We can also immediately return false if the string length exceeds the number of unique characters in the alphabet. After all, you can't form a string of 280 unique characters out of a 128-character alphabet.

It's also okay to assume 256 characters. This would be the case in extended ASCII. You should clarify your assumptions with your interviewer.
*/
